### DISCUSSION TITLE
Allow the resource :title attribute to reference a computed field in its resource.

### DISCUSSION THREAD

**Discussion by tvongaza** [avo-hq/avo#1549] [Feature requests]
A pretty standard use case is having a resource's title being made up of some computed value - example first & last names. From my testing it seems like the resource's title calls the backing model with the supplied title method. It would be nice to instead have it call its own resource's field, keeping all the resources display logic together.

Example of what I'd like to do:
```
class ContactResource < Avo::BaseResource
  self.title = :name
  self.includes = []

  field :id, as: :id
  field :name, as: :text, link_to_resource: true do |model|
    "#{model.first_name} #{model.last_name}"
  end
  field :first_name, as: :text, hide_on: [:index, :show]
  field :last_name, as: :text, hide_on: [:index, :show]
end
```

**Comment by Paul-Bob** [comment]
Hi @tvongaza. I can see several use cases here

## We want to use `name` on several lines of our app
In this scenario the correct approach would be to define a method inside the model:
```ruby
def name
  "#{first_name} #{last_name}"
end
```

## We want to use `name` only as a resource field (index and show views)
In this case we can do both, or we create the method inside the model as above (don't if you consider that's polluting the model with methods that just Avo uses) or we just can do that computed field as shown on your example.

## We want to use `name` only on title
In this case (right now) we are forced to create that method inside the model. This is not great. Here I see the limitation and I think making the resource's `title` attribute callable would solve the issue.  With `title` callable we would be able to do something like this.
```ruby
self.title = -> {  "#{model.first_name} #{model.last_name}" }
```

Why I think this is better than using the resource field? Because in this scenario we do not want that field only on title and that would force us to do something like this:
```ruby
class SomeResource < Avo::BaseResource
  self.title = :name

  # unused
  field :name, as: :text, hide_on: [:forms, :show, :index] do |model|
    "#{model.first_name} #{model.last_name}"
  end
end
```

## Described scenario where `name` it's used as field and as title
Now, if we go to the next scenario (that described by you) where you want to use it as a resource field and as a resource title you will end up with duplicated code, something like:
```ruby
class ContactResource < Avo::BaseResource
  self.title = -> { "#{model.first_name} #{model.last_name}" } # duplicated
  self.includes = []

  field :id, as: :id
  field :name, as: :text, link_to_resource: true do |model|
    "#{model.first_name} #{model.last_name}" # duplicated
  end
  field :first_name, as: :text, hide_on: [:index, :show]
  field :last_name, as: :text, hide_on: [:index, :show]
end
```

To solve this duplication i think we should have the following approach on `title` attribute:

1. Try to call `name` method on model.
2. If fails, try to get it from resource.
3. Any way allow it to be callable and computed without any of above dependencies.

@adrianthedev what do  you think?

**Reply by tvongaza** [reply]
Thanks @Paul-Bob, this came up because as I new user to Avo I expected the following code to use the `name` field on the resource, not call a `name` method on the backing model.

```
class SomeResource < Avo::BaseResource
  self.title = :name # Expected this to reference the :name field below, but this is not the case

  field :name, as: :text, hide_on: [:forms, :show, :index] do |model|
    "#{model.first_name} #{model.last_name}"
  end
end
```

I would suggest your title attribute order should be tweaked such that it tries to use the resource's field first, if not found tries to get it from the model.

Alternatively we could do something like:
```
self.title = -> { "#{resource.get_field:(name).value}" }  # or whatever the best way to get the computed value from the resource is - this just seems counter intuitive when in the mental context of its own file
```

**Reply by tvongaza** [reply]
I whipped up a quick MonkeyPatch to show how I as a new user expected the `self.title = :name` referencing the resources computed `:name` field to work. This is likely incomplete as I'm unsure everwhere in Avo the resource title is used.

```
module Aeon
  module Overrides
    module AvoBaseResource
      extend ActiveSupport::Concern
      included do
        def model_title
          return get_field(self.title).value
        end
      end
    end
    module AvoFieldsBelongsToField
      extend ActiveSupport::Concern
      included do
        def label
          self.target_resource.hydrate(model: self.value).model_title
        end
      end
    end
  end
end

Avo::BaseResource.include(Aeon::Overrides::AvoBaseResource)
Avo::Fields::BelongsToField.include(Aeon::Overrides::AvoFieldsBelongsToField)
```
